File: app/src/main/java/org/antrack/app/libs/Checks.java
package org.antrack.app.libs;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.Signature;
import android.util.Base64;
import android.util.Log;

import java.security.MessageDigest;

public class Checks {
    private static final String TAG = "Checks";
    private static final String SIGNATURE = "PN1RUozuVqArQa6drULZRbOErqI";
    private static final String PLAY_STORE_APP_ID = "com.android.vending";

    public static boolean all(Context context) {
        boolean signature = checkSignature(context);
        boolean fromPlayStore = checkInstaller(context);
        boolean isEmulator = checkEmulator();
        boolean isDebuggable = checkDebuggable(context);

        // DEBUG
        L.e(TAG, "signature: " + signature);
        L.e(TAG, "fromPlayStore: " + fromPlayStore);
        L.e(TAG, "isEmulator: " + isEmulator);
        L.e(TAG, "isDebuggable: " + isDebuggable);

        return signature &&
                fromPlayStore &&
                !isEmulator &&
                !isDebuggable;
    }

    public static boolean checkSignature(Context context) {
        return SIGNATURE.equals(getSignature(context));
    }

    public static String getSignature(Context context) {
        String apkSignature = null;
        try {
            PackageInfo packageInfo = context.getPackageManager()
                    .getPackageInfo(context.getPackageName(),
                            PackageManager.GET_SIGNATURES);

            for (Signature signature : packageInfo.signatures) {
                MessageDigest md = MessageDigest.getInstance("SHA");
                md.update(signature.toByteArray());
                apkSignature = Base64.encodeToString(md.digest(), Base64.DEFAULT);
                L.e("DEBUG", "SIGNATURE: " + apkSignature);
            }
        } catch (Exception e) {}
        return apkSignature;
    }

    public static boolean checkInstaller(Context context) {
            final String installer = context.getPackageManager()
                    .getInstallerPackageName(context.getPackageName());

            return installer != null
                    && installer.startsWith(PLAY_STORE_APP_ID);

    }

    public static boolean checkEmulator() {
        try {
            boolean goldfish = getSystemProperty("ro.hardware").contains("goldfish");
            boolean emu = getSystemProperty("ro.kernel.qemu").length() > 0;
            boolean sdk = getSystemProperty("ro.product.model").equals("sdk");

            if (emu || goldfish || sdk) {
                return true;
            }
        } catch (Exception e) {}
        return false;
    }

    public static boolean checkDebuggable(Context context){
        return (context.getApplicationInfo().flags & ApplicationInfo.FLAG_DEBUGGABLE) != 0;

    }

    private static String getSystemProperty(String name) throws Exception {
        Class systemPropertyClazz = Class.forName("android.os.SystemProperties");
        return (String) systemPropertyClazz.getMethod("get", new Class[]{String.class})
                .invoke(systemPropertyClazz, new Object[]{name});
    }
}

File: app/src/main/java/org/antrack/app/libs/Checks.kt
package org.antrack.app.libs

import android.content.Context
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.util.Base64
import java.security.MessageDigest

object Checks {
    private val TAG = "Checks"
    private val SIGNATURE = "PN1RUozuVqArQa6drULZRbOErqI="
    private val PLAY_STORE_APP_ID = "com.android.vending"

    fun all(context: Context): Boolean {
        val signature = checkSignature(context)
        val fromPlayStore = checkInstaller(context)
        val isEmulator = checkEmulator()
        val isDebuggable = checkDebuggable(context)

        // DEBUG
        L.e(TAG, "signature: " + signature)
        L.e(TAG, "fromPlayStore: " + fromPlayStore)
        L.e(TAG, "isEmulator: " + isEmulator)
        L.e(TAG, "isDebuggable: " + isDebuggable)

        return signature &&
                fromPlayStore &&
                !isEmulator &&
                !isDebuggable
    }

    fun checkSignature(context: Context): Boolean {
        return SIGNATURE == getSignature(context)!!.trim { it <= ' ' }
    }

    fun getSignature(context: Context): String? {
        var apkSignature: String? = null
        try {
            val packageInfo = context.packageManager
                    .getPackageInfo(context.packageName,
                            PackageManager.GET_SIGNATURES)

            for (signature in packageInfo.signatures) {
                val md = MessageDigest.getInstance("SHA")
                md.update(signature.toByteArray())
                apkSignature = Base64.encodeToString(md.digest(), Base64.DEFAULT)
                L.e("DEBUG", "SIGNATURE: " + apkSignature!!)
            }
        } catch (e: Exception) {
        }

        return apkSignature
    }

    fun checkInstaller(context: Context): Boolean {
        val installer = context.packageManager
                .getInstallerPackageName(context.packageName)

        return installer != null && installer.startsWith(PLAY_STORE_APP_ID)

    }

    fun checkEmulator(): Boolean {
        try {
            val goldfish = getSystemProperty("ro.hardware").contains("goldfish")
            val emu = getSystemProperty("ro.kernel.qemu").length > 0
            val sdk = getSystemProperty("ro.product.model") == "sdk"

            if (emu || goldfish || sdk) {
                return true
            }
        } catch (e: Exception) {
        }

        return false
    }

    fun checkDebuggable(context: Context): Boolean {
        return context.applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE != 0

    }

    @Throws(Exception::class)
    private fun getSystemProperty(name: String): String {
        val systemPropertyClazz = Class.forName("android.os.SystemProperties")
        return systemPropertyClazz.getMethod("get", *arrayOf<Class<*>>(String::class.java))
                .invoke(systemPropertyClazz, *arrayOf<Any>(name)) as String
    }
}
